### ISSUE TITLE
HMR only works for modules that do not contain function calls at the top level

### ISSUE THREAD

**Issue by nicolo-ribaudo** [denoland/deno#21118]
main.js:
```js
import "./dep.js";
console.log("Running main");
setInterval(() => {}, 1000); // keep alive
```

dep.js:
```js
let x = Number("2")

export function fn() {
  return 3;
}
```

Run with `deno run --unstable-hmr main.js`.

If you edit the return value of `fn` (for example, replacing `3` with `4`), it will re-run `main.js`. The reason given is "blocked by top-level ES module change", but I'm not changing anything at the top level of the module.

If I use `2` instead of `Number("2")`, the problem goes away.

_This issue may be a duplicate of https://github.com/denoland/deno/issues/21101, but I'm not sure about what's the root cause there._

**Comment by bartlomieju** [comment]
Yeah, it's a bit surprising, but that's pretty much what V8 gives us: https://chromedevtools.github.io/devtools-protocol/tot/Debugger/#method-setScriptSource

**Comment by nicolo-ribaudo** [comment]
I'm not sure about what I saw when I reported this issue, but it looks like HMR doesn't work if there is _anything_ other than a single function declaration? 🤔 

Even with these two `dep.js` files when I try to change the return value it reloads `main.js`:
```js
0;

export function fn() {
  return 8;
}
```

```js
function f() {}

export function fn() {
  return 8;
}
```

**Comment by nicolo-ribaudo** [comment]
Also, note that in Chrome devtools this works. I slightly modified the `main.js` in the original script to log the return value of `fn()`:

```js
// main.js
import * as d from "./dep.js"
console.log("Running main");
setInterval(() => console.log("ping", d.fn()), 1000);

// dep.js
let x = Number("2");
export function fn() {
  return 10;
}
```

When running in Deno, if I edit the return value of `fn()` it re-loads also `main.js`:

https://github.com/denoland/deno/assets/7000710/eb544563-6006-4ae3-ba4b-61093523fbc8

In Chrome, it does not re-run `main.js`:

https://github.com/denoland/deno/assets/7000710/9a1a6826-4284-48ca-88f6-e61a414e5d64

**Comment by bartlomieju** [comment]
I'm wondering if the problem here is because of the source maps being embedded at the end of the file - when a file changes we transpile and not append the source map (because that always caused full restart). But you are using `.js` files which are not transpiled at all and shouldn't include source maps 🤔

**Comment by bartlomieju** [comment]
@devsnek can you help me debug this one?

**Comment by devsnek** [comment]
This seems to have been fixed in one of the V8 upgrades.
